Reqnroll.VisualStudio/Editor/Services/UnusedStepDefinitionsFinder.cs

```diff
@@ -0,0 +1,127 @@
+
+public class UnusedStepDefinitionsFinder
+{
+    private readonly IIdeScope _ideScope;
+
+    public UnusedStepDefinitionsFinder(IIdeScope ideScope)
+    {
+        _ideScope = ideScope;
+    }
+
+    public IEnumerable<ProjectStepDefinitionBinding> FindUnused(ProjectStepDefinitionBinding[] stepDefinitions,
+                                                        string[] featureFiles, DeveroomConfiguration configuration)
+    {
+        return featureFiles.SelectMany(ff => FindUnused(stepDefinitions, ff, configuration));
+    }
+
+    public IEnumerable<ProjectStepDefinitionBinding> FindUnused(ProjectStepDefinitionBinding[] stepDefinitions, 
+        string featureFilePath, DeveroomConfiguration configuration) => 
+        LoadContent(featureFilePath, out string featureFileContent)
+        ? FindUsagesFromContent(stepDefinitions, featureFileContent, featureFilePath, configuration)
+        : Enumerable.Empty<ProjectStepDefinitionBinding>();
+
+    private IEnumerable<ProjectStepDefinitionBinding> FindUsagesFromContent(ProjectStepDefinitionBinding[] stepDefinitions, string featureFileContent, string featureFilePath, DeveroomConfiguration configuration)
+    {
+        var dialectProvider = ReqnrollGherkinDialectProvider.Get(configuration.DefaultFeatureLanguage);
+        var parser = new DeveroomGherkinParser(dialectProvider, _ideScope.MonitoringService);
+        parser.ParseAndCollectErrors(featureFileContent, _ideScope.Logger,
+            out var gherkinDocument, out _);
+
+        var featureNode = gherkinDocument?.Feature;
+        if (featureNode == null)
+            return Enumerable.Empty<ProjectStepDefinitionBinding>();
+
+        var dummyRegistry = ProjectBindingRegistry.FromBindings(stepDefinitions);
+
+        //this set keeps track of which step definitions have not been used. We start with ALL step definitions, and will subtract those that match steps in the feature file
+        HashSet<ProjectStepDefinitionBinding> stepDefinitionsSet = new(stepDefinitions);
+
+        var featureContext = new UnusedFinderContext(featureNode);
+
+        foreach (var scenarioDefinition in featureNode.FlattenStepsContainers())
+        {
+            var context = new UnusedFinderContext(scenarioDefinition, featureContext);
+
+            foreach (var step in scenarioDefinition.Steps)
+            {
+                var matchResult = dummyRegistry.MatchStep(step, context);
+
+                var usedSteps = matchResult.Items.Where(m => m.Type == MatchResultType.Defined || m.Type == MatchResultType.Ambiguous)
+                                                 .Select(i => i.MatchedStepDefinition)
+                                                 .ToHashSet<ProjectStepDefinitionBinding>();
+                stepDefinitionsSet.ExceptWith(usedSteps);
+            }
+        }
+
+        return stepDefinitionsSet.ToList();
+
+    }
+
+    //TODO: duplicate of what is in StepDefinitionUsageFinder; consider refactoring to shared implementation
```

Comment:
Agree. Maybe a `FeatureFileVisitor` or `FeatureFileProcessor` abstract base class?

Reply:
Created a base class for both Finder services, called it StepFinderBase.cs. IT contains the common logic for loading feature file content.